proxy/src/telemetry/metrics/mod.rs

```diff
@@ -207,10 +224,123 @@ impl fmt::Display for Root {
     }
 }
 
+// ===== impl Stamped =====
+
+impl<T> Stamped<T> {
+    fn stamped(&mut self) -> &mut T {
+        self.stamp = Instant::now();
+        &mut self.inner
+    }
+}
+
+impl<T> From<T> for Stamped<T> {
+    fn from(inner: T) -> Self {
+        Self {
+            inner,
+            stamp: Instant::now(),
+        }
+    }
+}
+
+impl<T: Default> Default for Stamped<T> {
+    fn default() -> Self {
+        T::default().into()
+    }
+}
+
+impl<T> ::std::ops::Deref for Stamped<T> {
```

Comment:
I've been told multiple times that `Deref` conversions (i.e. "implemeting `Deref` for things that aren't smart pointers") are "dangerous" and "considered harmful". 

With that said, I subscribe to a definition of "smart pointer" broad enough that this could be considered a smart pointer, and it isn't part of public API, so I think this is a valid use, but others may not agree.